File: frontend/src/app/features/tasks/services/task.service.ts
import { HttpClient, HttpErrorResponse, HttpParams } from '@angular/common/http';
import { Injectable, inject, signal } from '@angular/core';
import { Observable, catchError, firstValueFrom, throwError } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { Task, TaskCreate, TaskFilters, TaskList, TaskUpdate } from '../../../models/task.model';

/**
 * Service for task-related operations.
 * Provides both Observable and Promise-based (async/await) methods.
 */
@Injectable({
  providedIn: 'root'
})
export class TaskService {
  private http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiBaseUrl}/tasks`;

  // Signals for state management
  private isLoadingSignal = signal<boolean>(false);
  private errorSignal = signal<string | null>(null);

  // Public readonly signals
  isLoading = this.isLoadingSignal.asReadonly();
  error = this.errorSignal.asReadonly();

  /**
   * List tasks with optional filters and pagination - Observable version
   */
  listTasks(
    page: number = 1,
    pageSize: number = 20,
    filters?: TaskFilters
  ): Observable<TaskList> {
    this.errorSignal.set(null);
    const params = this.buildQueryParams(page, pageSize, filters);
    return this.http.get<TaskList>(this.apiUrl, { params })
      .pipe(catchError(this.handleError.bind(this)));
  }

  /**
   * List tasks with optional filters and pagination - Async version
   */
  async listTasksAsync(
    page: number = 1,
    pageSize: number = 20,
    filters?: TaskFilters
  ): Promise<TaskList> {
    this.errorSignal.set(null);
    this.isLoadingSignal.set(true);

    try {
      return await firstValueFrom(this.listTasks(page, pageSize, filters));
    } finally {
      this.isLoadingSignal.set(false);
    }
  }

  /**
   * Get a specific task by ID - Observable version
   */
  getTask(taskId: string): Observable<Task> {
    this.errorSignal.set(null);
    return this.http.get<Task>(`${this.apiUrl}/${taskId}`)
      .pipe(catchError(this.handleError.bind(this)));
  }

  /**
   * Get a specific task by ID - Async version
   */
  async getTaskAsync(taskId: string): Promise<Task> {
    this.errorSignal.set(null);
    this.isLoadingSignal.set(true);

    try {
      return await firstValueFrom(this.getTask(taskId));
    } finally {
      this.isLoadingSignal.set(false);
    }
  }

  /**
   * Create a new task - Observable version
   */
  createTask(task: TaskCreate): Observable<Task> {
    this.errorSignal.set(null);
    return this.http.post<Task>(this.apiUrl, task)
      .pipe(catchError(this.handleError.bind(this)));
  }

  /**
   * Create a new task - Async version
   */
  async createTaskAsync(task: TaskCreate): Promise<Task> {
    this.errorSignal.set(null);
    this.isLoadingSignal.set(true);

    try {
      return await firstValueFrom(this.createTask(task));
    } finally {
      this.isLoadingSignal.set(false);
    }
  }

  /**
   * Update an existing task - Observable version
   */
  updateTask(taskId: string, task: TaskUpdate): Observable<Task> {
    this.errorSignal.set(null);
    return this.http.put<Task>(`${this.apiUrl}/${taskId}`, task)
      .pipe(catchError(this.handleError.bind(this)));
  }

  /**
   * Update an existing task - Async version
   */
  async updateTaskAsync(taskId: string, task: TaskUpdate): Promise<Task> {
    this.errorSignal.set(null);
    this.isLoadingSignal.set(true);

    try {
      return await firstValueFrom(this.updateTask(taskId, task));
    } finally {
      this.isLoadingSignal.set(false);
    }
  }

  /**
   * Delete a task - Observable version
   */
  deleteTask(taskId: string): Observable<void> {
    this.errorSignal.set(null);
    return this.http.delete<void>(`${this.apiUrl}/${taskId}`)
      .pipe(catchError(this.handleError.bind(this)));
  }

  /**
   * Delete a task - Async version
   */
  async deleteTaskAsync(taskId: string): Promise<void> {
    this.errorSignal.set(null);
    this.isLoadingSignal.set(true);

    try {
      await firstValueFrom(this.deleteTask(taskId));
    } finally {
      this.isLoadingSignal.set(false);
    }
  }

  /**
   * Mark a task as completed - Observable version
   */
  completeTask(taskId: string): Observable<Task> {
    this.errorSignal.set(null);
    return this.http.patch<Task>(`${this.apiUrl}/${taskId}/complete`, {})
      .pipe(catchError(this.handleError.bind(this)));
  }

  /**
   * Mark a task as completed - Async version
   */
  async completeTaskAsync(taskId: string): Promise<Task> {
    this.errorSignal.set(null);
    this.isLoadingSignal.set(true);

    try {
      return await firstValueFrom(this.completeTask(taskId));
    } finally {
      this.isLoadingSignal.set(false);
    }
  }

  /**
   * Reopen a completed task - Observable version
   */
  reopenTask(taskId: string): Observable<Task> {
    this.errorSignal.set(null);
    return this.http.patch<Task>(`${this.apiUrl}/${taskId}/reopen`, {})
      .pipe(catchError(this.handleError.bind(this)));
  }

  /**
   * Reopen a completed task - Async version
   */
  async reopenTaskAsync(taskId: string): Promise<Task> {
    this.errorSignal.set(null);
    this.isLoadingSignal.set(true);

    try {
      return await firstValueFrom(this.reopenTask(taskId));
    } finally {
      this.isLoadingSignal.set(false);
    }
  }

  /**
   * Clear error message
   */
  clearError(): void {
    this.errorSignal.set(null);
  }

  /**
   * Build query parameters for list tasks request
   */
  private buildQueryParams(
    page: number,
    pageSize: number,
    filters?: TaskFilters
  ): HttpParams {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('page_size', pageSize.toString());

    if (filters) {
      if (filters.status) {
        params = params.set('status', filters.status);
      }
      if (filters.priority) {
        params = params.set('priority', filters.priority);
      }
      if (filters.search) {
        params = params.set('search', filters.search);
      }
      if (filters.sort_by) {
        params = params.set('sort_by', filters.sort_by);
      }
      if (filters.tag_ids && filters.tag_ids.length > 0) {
        filters.tag_ids.forEach(tagId => {
          params = params.append('tag_ids', tagId);
        });
      }
    }

    return params;
  }

  /**
   * Handle HTTP errors
   */
  private handleError(error: HttpErrorResponse): Observable<never> {
    let errorMessage = 'An unknown error occurred';

    if (error.error instanceof ErrorEvent) {
      // Client-side or network error
      errorMessage = `Error: ${error.error.message}`;
    } else {
      // Backend returned an unsuccessful response code
      errorMessage = error.error?.detail || error.error?.message || error.message;
    }

    this.errorSignal.set(errorMessage);
    console.error('Task Service Error:', errorMessage);
    return throwError(() => error);
  }
}

File: backend/app/services/task_service.py
"""
Task service with business logic.
"""
from typing import List, Optional
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncSession

from app.api.schemas.task import TaskCreate, TaskList, TaskOut, TaskUpdate
from app.db.models.task import TaskStatus
from app.db.repositories.task_repository import TaskRepository


class TaskService:
    """Service for task-related business logic."""

    def __init__(self, db: AsyncSession):
        self.db = db
        self.repository = TaskRepository(db)

    async def list_tasks(
        self,
        user_id: UUID,
        page: int = 1,
        page_size: int = 20,
        status: Optional[TaskStatus] = None,
        priority: Optional[str] = None,
        tag_ids: Optional[List[UUID]] = None,
        search: Optional[str] = None,
        sort_by: str = "created_at"
    ) -> TaskList:
        """List tasks with filtering, search, and pagination."""
        skip = (page - 1) * page_size

        tasks = await self.repository.list_by_user(
            user_id=user_id,
            skip=skip,
            limit=page_size,
            status=status,
            priority=priority,
            tag_ids=tag_ids,
            search=search,
            sort_by=sort_by
        )

        total = await self.repository.count_by_user(
            user_id=user_id,
            status=status,
            priority=priority,
            tag_ids=tag_ids,
            search=search
        )

        return TaskList(
            items=[TaskOut.model_validate(task) for task in tasks],
            total=total,
            page=page,
            page_size=page_size,
            pages=(total + page_size - 1) // page_size
        )

    async def create_task(
        self,
        user_id: UUID,
        task_in: TaskCreate
    ) -> TaskOut:
        """Create a new task."""
        task_data = task_in.model_dump(exclude={"tag_ids"})
        task_data["user_id"] = user_id

        task = await self.repository.create(task_data)

        # Associate tags if provided
        if task_in.tag_ids:
            await self.repository.associate_tags(task, task_in.tag_ids)

        await self.db.refresh(task, ["tags"])
        return TaskOut.model_validate(task)

    async def get_task(
        self,
        task_id: UUID,
        user_id: UUID
    ) -> Optional[TaskOut]:
        """Get a task by ID."""
        task = await self.repository.get_by_user(user_id, task_id)
        return TaskOut.model_validate(task) if task else None

    async def update_task(
        self,
        task_id: UUID,
        user_id: UUID,
        task_in: TaskUpdate
    ) -> Optional[TaskOut]:
        """Update a task."""
        task = await self.repository.get_by_user(user_id, task_id)
        if not task:
            return None

        update_data = task_in.model_dump(exclude_unset=True, exclude={"tag_ids"})

        for field, value in update_data.items():
            if hasattr(task, field):
                setattr(task, field, value)

        # Update tags if provided
        if task_in.tag_ids is not None:
            await self.repository.associate_tags(task, task_in.tag_ids)

        await self.db.flush()
        await self.db.refresh(task, ["tags"])
        return TaskOut.model_validate(task)

    async def delete_task(
        self,
        task_id: UUID,
        user_id: UUID
    ) -> bool:
        """Soft delete a task."""
        return await self.repository.soft_delete(task_id, user_id)

    async def complete_task(
        self,
        task_id: UUID,
        user_id: UUID
    ) -> Optional[TaskOut]:
        """Mark a task as completed."""
        task = await self.repository.complete_task(task_id, user_id)
        return TaskOut.model_validate(task) if task else None

    async def reopen_task(
        self,
        task_id: UUID,
        user_id: UUID
    ) -> Optional[TaskOut]:
        """Reopen a completed task."""
        task = await self.repository.reopen_task(task_id, user_id)
        return TaskOut.model_validate(task) if task else None
